gcp-deployer/deploy/deploy.go

```diff
@@ -73,12 +75,14 @@ func NewDeployer(provider string, kubeConfigPath string, machineSetupConfigPath
 }
 
 func (d *deployer) CreateCluster(c *clusterv1.Cluster, machines []*clusterv1.Machine) error {
+	createSshKeyPairs()
```

Comment:
There is a lot of logic for managing the ssh key files, and reading / writing them (for example in gcp-deployer/deploy/ssh.go). In this function, there is complexity around calling createSshKeyPairs(...) and then calling cleanupSshKeyPairs(...) at the right time. createSshKeyPairs(...) is also called in NewDeployer and it isn't immediately obvious why. While this could probably be simplified by adding "defer cleanupSshKeyPairs(...)" below "createSshKeyPairs(...)". It seems unfortunate that the deployer has to know how to manage the keys. 

I think it would be simpler if machineactuator.go and the deployer simply dealt with keys that were stored in memory. They could be loaded once and then reused. The logic that decides whether or not to generate keys could be kicked off once. Perhaps it could work like 

if !userSuppliesKeys {
  // kick off key generation
}
// happy path below, everything makes use of the keys that are stored in memory and available.

From a security perspective, if that is a concern, the program currently has access to read the key files so any malicious user that has the access to dump the program's memory must be able to access the key files. In addition, we are storing the keys in memory, even if temporarily, while we use them.

Reply:
I agree that It would be nice if things were in memory. There are challenges to that in this situation:
- In the deployer, the ssh keys need to persist between runs of different deployer commands. Once the ssh keys are used for creating a cluster, we need the keys around till that cluster is destroyed.
- For the machine controller, if I want it to be able to work even if the apiserver is down, the keys need to be passed in via environment variable or as a file in command line arguments, file seemed the cleanest/safest.

Reply:
Yes, it is totally reasonable that the ssh keys would be persisted to disk and reloaded with each instance of a process that needs tehm. However, what I was trying to convey is maybe we shouldn't reload them on every action: CreateCluster, Update/AddMachine, etc. However, the current version of the change is simpler.

One thing you may consider is a design that looks something like this (maybe this is the wrong way to go though):

In ssh.go,
```
type Ssh struct {
 // stuff needed to save for ssh
  publicKey string,
  privateKey string,
}

func New()(*Ssh, error)  {
  // generate keys here
  // ... more code for generation
  return NewWithKeys(generatedPrivateKey, generatedPublicKey)
}

func NewWithKeys(privateKey, publicKey)(*Ssh, error)  {
   //
   return NewWithKeysAndUser(privateKey, publicKey, defaultUser)
}

func NewWithKeysAndUser(privateKey, publicKey, userName) (*Ssh, error) {
  return new &Ssh {
     // set variables here
  }, nil


func (ssh *Ssh) Command(ip string, cmd string) (string, error) {
   // use the 'ssh' variable here
}
```

Another developer would use it as follows (to run ls -al in this case):

```
ssh := ssh.New(privateKeyPath, publicKeyPath, user)
ssh.Command(ip, "ls -al")
```

And there could be convienence methods for creating a New SSH from files as well. In that way the rest of the application would only need to pass around a single Ssh struct instead of the privateKeyPath, publicKeyPath, and user.

Reply:
Switched to pre-loading public key. private key needs to stay as file since ssh only takes a file for that param (understandably as private keys are very sensitive).